### ISSUE TITLE
Unexpected errors are reported for local functions declared within a switch statement.

### ISSUE THREAD

**Issue by AlekseyTs** [dotnet/roslyn#13029]
```
        [Fact]
        public void ScopeOfLocalFunction()
        {
            var source =
@"
public class X
{
    public static void Main()
    {
    }

    bool Dummy(params object[] x) { return true; }

    void Test14(int val)
    {
        switch (val)
        {
            case 1 when TakeOutParam(true, out var x14):
                void x14() {return;};
                break;
            case 2:
                x14();
                break;
        }

        switch (val)
        {
            case 1 when Dummy(1 is var x14):
                void x14() {return;};
                break;
            case 2:
                x14();
                break;
        }
    }

    static bool TakeOutParam<T>(T y, out T x) 
    {
        x = y;
        return true;
    }
}
";
            var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugExe, parseOptions: TestOptions.Regular);

            compilation.VerifyDiagnostics(
                // (14,52): error CS0136: A local or parameter named 'x14' cannot be declared in this scope because that name is used in an enclosing local scope to define a local or parameter
                //             case 1 when TakeOutParam(true, out var x14):
                Diagnostic(ErrorCode.ERR_LocalIllegallyOverrides, "x14").WithArguments("x14").WithLocation(14, 52),
                // (24,40): error CS0136: A local or parameter named 'x14' cannot be declared in this scope because that name is used in an enclosing local scope to define a local or parameter
                //             case 1 when Dummy(1 is var x14):
                Diagnostic(ErrorCode.ERR_LocalIllegallyOverrides, "x14").WithArguments("x14").WithLocation(24, 40)
                );
        }
```

Unexpected errors reported:

```
                // (15,22): error CS0136: A local or parameter named 'x14' cannot be declared in this scope because that name is used in an enclosing local scope to define a local or parameter
                //                 void x14() {return;};
                Diagnostic(ErrorCode.ERR_LocalIllegallyOverrides, "x14").WithArguments("x14").WithLocation(15, 22),
                // (25,22): error CS0136: A local or parameter named 'x14' cannot be declared in this scope because that name is used in an enclosing local scope to define a local or parameter
                //                 void x14() {return;};
                Diagnostic(ErrorCode.ERR_LocalIllegallyOverrides, "x14").WithArguments("x14").WithLocation(25, 22),
                // (18,17): error CS0165: Use of unassigned local variable 'x14'
                //                 x14();
                Diagnostic(ErrorCode.ERR_UseDefViolation, "x14()").WithArguments("x14").WithLocation(18, 17),
                // (28,17): error CS0165: Use of unassigned local variable 'x14'
                //                 x14();
                Diagnostic(ErrorCode.ERR_UseDefViolation, "x14()").WithArguments("x14").WithLocation(28, 17)
```

**Comment by gafter** [comment]
Note that replacing

```cs
                void x14() { return; };
```

with

```cs
                Action x14 = null;
```

illustrates the problem specific to local functions.

The definite assignment errors are expected and correct with the current specification and implementation, but that is expected to change.
